File: sdks/python/temps_plugin/runtime.py
# SPDX-FileCopyrightText: 2024-2026 Temps Contributors
# SPDX-License-Identifier: MIT OR Apache-2.0

"""Plugin runtime — handles the full lifecycle of a Temps external plugin.

Mirrors: sdks/node/packages/plugin-sdk/src/runtime.ts

Lifecycle:
1. Parse CLI arguments (--socket-path, --auth-secret, --data-dir)
2. Emit manifest to stdout (handshake phase 1)
3. Start aiohttp server on Unix domain socket
4. Emit ready to stdout (handshake phase 2)
5. Accept WebSocket connection from host on /_temps/channel
6. Create PluginContext with TempsClient
7. Call plugin.on_start()
8. Serve plugin routes, health, events, and embedded UI
9. Handle SIGTERM for graceful shutdown
"""

from __future__ import annotations

import asyncio
import json
import os
import signal
import sys
from pathlib import Path
from typing import Any, Awaitable, Callable, Protocol

from aiohttp import web

from temps_plugin.client import TempsClient
from temps_plugin.context import PluginContext
from temps_plugin.protocol import (
    PLUGIN_CHANNEL_PATH,
    emit_manifest,
    emit_ready,
    extract_auth_context,
)
from temps_plugin.types import (
    EmbeddedAssets,
    PluginArgs,
    PluginEvent,
    PluginManifest,
)
from temps_plugin.ui import serve_embedded_ui

# Type aliases for plugin callbacks
RequestHandler = Callable[[web.Request], Awaitable[web.Response]]


class TempsPlugin(Protocol):
    """Protocol that plugin implementations must satisfy."""

    def manifest(self) -> PluginManifest: ...
    def handler(self, ctx: PluginContext) -> RequestHandler: ...
    def embedded_ui_assets(self) -> EmbeddedAssets | None: ...

    async def on_start(self, ctx: PluginContext) -> None: ...
    def on_event(self, ctx: PluginContext, event: PluginEvent) -> None: ...
    async def on_shutdown(self) -> None: ...


def run_plugin(plugin: TempsPlugin) -> None:
    """Run a Temps plugin. This is the main entry point."""
    asyncio.run(_run_plugin_async(plugin))


async def _run_plugin_async(plugin: TempsPlugin) -> None:
    # 1. Parse CLI arguments
    args = _parse_args(sys.argv[1:])
    os.makedirs(args.data_dir, exist_ok=True)

    manifest = plugin.manifest()

    # 2. Emit manifest (handshake phase 1)
    emit_manifest(manifest)

    # Ensure socket directory exists and clean up stale socket
    socket_dir = str(Path(args.socket_path).parent)
    os.makedirs(socket_dir, exist_ok=True)
    if os.path.exists(args.socket_path):
        os.unlink(args.socket_path)

    # Channel connection tracking
    channel_connected: asyncio.Future[TempsClient] = (
        asyncio.get_event_loop().create_future()
    )
    context: PluginContext | None = None
    plugin_ready = False

    # Embedded UI
    embedded_assets = plugin.embedded_ui_assets()
    has_ui = embedded_assets is not None

    # Plugin request handler (set after initialization)
    plugin_handler: RequestHandler | None = None

    # -- aiohttp request handlers --

    async def handle_health(_request: web.Request) -> web.Response:
        return web.json_response({"status": "ok", "plugin": manifest.name})

    async def handle_ws_channel(request: web.Request) -> web.WebSocketResponse:
        ws = web.WebSocketResponse()
        await ws.prepare(request)

        client = TempsClient(ws)
        client.start_reader()

        if not channel_connected.done():
            channel_connected.set_result(client)

        # Keep the WebSocket open until it closes
        await client._reader_task
        return ws

    async def handle_events(request: web.Request) -> web.Response:
        try:
            data = await request.json()
            event = PluginEvent(
                id=data.get("id", ""),
                event_type=data.get("event_type", ""),
                timestamp=data.get("timestamp", ""),
                project_id=data.get("project_id"),
                data=data.get("data", {}),
            )
            if context:
                plugin.on_event(context, event)
            return web.json_response({"ok": True})
        except Exception as e:
            return web.json_response(
                {"error": "Invalid event payload", "detail": str(e)},
                status=400,
            )

    async def handle_ui(request: web.Request) -> web.Response:
        if embedded_assets is None:
            return web.Response(text="Not Found", status=404)
        return serve_embedded_ui(request.path, embedded_assets)

    async def handle_catch_all(request: web.Request) -> web.Response:
        path = request.path

        # UI redirect
        if path == "/ui":
            raise web.HTTPFound("/ui/")

        # Embedded UI
        if path.startswith("/ui/") and embedded_assets:
            return serve_embedded_ui(path, embedded_assets)

        # Not initialized yet
        if not plugin_ready or plugin_handler is None:
            return web.Response(text="Plugin initializing", status=503)

        # Delegate to plugin handler
        return await plugin_handler(request)

    # 3. Build and start the aiohttp app
    app = web.Application()
    app.router.add_get(manifest.health_path, handle_health)
    if manifest.health_path != "/health":
        app.router.add_get("/health", handle_health)
    app.router.add_get(PLUGIN_CHANNEL_PATH, handle_ws_channel)
    app.router.add_post("/_events", handle_events)
    # Catch-all for plugin routes and UI
    app.router.add_route("*", "/{path_info:.*}", handle_catch_all)

    runner = web.AppRunner(app)
    await runner.setup()
    site = web.UnixSite(runner, args.socket_path)
    await site.start()

    # 4. Emit ready (handshake phase 2)
    emit_ready(has_ui)

    # Set up signal handlers early so SIGTERM works during any phase
    shutdown_event = asyncio.Event()

    def _signal_handler() -> None:
        shutdown_event.set()

    loop = asyncio.get_event_loop()
    for sig in (signal.SIGTERM, signal.SIGINT):
        loop.add_signal_handler(sig, _signal_handler)

    # 5. Wait for the host to connect via WebSocket channel
    # If SIGTERM arrives during this wait, we abort and shut down.
    client: TempsClient | None = None
    try:
        channel_or_shutdown = asyncio.ensure_future(channel_connected)
        shutdown_task = asyncio.create_task(shutdown_event.wait())
        done, _ = await asyncio.wait(
            {channel_or_shutdown, shutdown_task},
            timeout=30.0,
            return_when=asyncio.FIRST_COMPLETED,
        )
        if shutdown_task in done:
            # SIGTERM received during channel wait
            channel_or_shutdown.cancel()
            await _shutdown(plugin, manifest, None, runner, args)
            return
        if channel_or_shutdown in done:
            client = channel_or_shutdown.result()
        else:
            # Timeout — no host connected
            channel_or_shutdown.cancel()
            shutdown_task.cancel()
            print(
                "[temps-plugin] Warning: platform channel not connected, "
                "running without platform data access",
                file=sys.stderr,
            )
    except Exception:
        print(
            "[temps-plugin] Warning: platform channel not connected, "
            "running without platform data access",
            file=sys.stderr,
        )

    # 6. Create PluginContext
    ctx = PluginContext(
        plugin_name=manifest.name,
        data_dir=args.data_dir,
        auth_secret=args.auth_secret,
        client=client,
    )
    context = ctx

    if client:
        client.on_event(lambda event: plugin.on_event(ctx, event))

    # 7. Call plugin.on_start()
    try:
        await plugin.on_start(ctx)
    except Exception as e:
        print(f"[temps-plugin] Error in on_start: {e}", file=sys.stderr)
        raise

    # 8. Mount the plugin handler
    plugin_handler = plugin.handler(ctx)
    plugin_ready = True

    print(
        f"[temps-plugin] {manifest.display_name or manifest.name} "
        f"v{manifest.version} running on {args.socket_path}",
        file=sys.stderr,
    )

    # 9. Graceful shutdown -- wait until signal
    await shutdown_event.wait()

    await _shutdown(plugin, manifest, client, runner, args)


async def _shutdown(
    plugin: TempsPlugin,
    manifest: PluginManifest,
    client: TempsClient | None,
    runner: web.AppRunner,
    args: PluginArgs,
) -> None:
    """Run graceful shutdown sequence."""
    print(f"[temps-plugin] Shutting down {manifest.name}...", file=sys.stderr)
    try:
        await plugin.on_shutdown()
    except Exception as e:
        print(f"[temps-plugin] Error during shutdown: {e}", file=sys.stderr)

    if client:
        await client.close()

    await runner.cleanup()

    # Clean up socket file
    if os.path.exists(args.socket_path):
        try:
            os.unlink(args.socket_path)
        except OSError:
            pass


def _parse_args(argv: list[str]) -> PluginArgs:
    """Parse CLI arguments."""
    socket_path: str | None = None
    auth_secret: str | None = None
    data_dir: str | None = None
    database_url: str | None = None

    i = 0
    while i < len(argv):
        arg = argv[i]
        next_val = argv[i + 1] if i + 1 < len(argv) else None

        if arg == "--socket-path" and next_val:
            socket_path = next_val
            i += 2
        elif arg == "--auth-secret" and next_val:
            auth_secret = next_val
            i += 2
        elif arg == "--data-dir" and next_val:
            data_dir = next_val
            i += 2
        elif arg == "--database-url" and next_val:
            database_url = next_val
            i += 2
        else:
            i += 1

    if not socket_path:
        raise SystemExit("Error: --socket-path is required")
    if not auth_secret:
        raise SystemExit("Error: --auth-secret is required")
    if not data_dir:
        raise SystemExit("Error: --data-dir is required")

    return PluginArgs(
        socket_path=socket_path,
        auth_secret=auth_secret,
        data_dir=data_dir,
        database_url=database_url,
    )

File: sdks/node/packages/plugin-sdk/src/runtime.ts
// SPDX-FileCopyrightText: 2024-2026 Temps Contributors
// SPDX-License-Identifier: MIT OR Apache-2.0

/**
 * Plugin runtime -- handles the full lifecycle of a Temps external plugin.
 *
 * Mirrors: crates/temps-plugin-sdk/src/runtime.rs
 *
 * Built for Bun -- compiled to a single binary via `bun build --compile`.
 *
 * Lifecycle:
 * 1. Parse CLI arguments (--socket-path, --auth-secret, --data-dir, etc.)
 * 2. Emit manifest to stdout (handshake phase 1)
 * 3. Start Bun.serve on Unix domain socket
 * 4. Emit ready to stdout (handshake phase 2)
 * 5. Accept WebSocket connection from host on /_temps/channel
 * 6. Create PluginContext with TempsClient
 * 7. Call plugin.onStart()
 * 8. Serve plugin routes, health, events, and embedded UI
 * 9. Handle SIGTERM for graceful shutdown
 */

import { existsSync, mkdirSync, unlinkSync } from "node:fs";
import { dirname } from "node:path";
import type { TempsPlugin, PluginArgs, PluginEvent } from "./types.js";
import { TempsClient, type WsLike } from "./client.js";
import { PluginContext } from "./context.js";
import { emitManifest, emitReady, PLUGIN_CHANNEL_PATH, extractAuthContext } from "./protocol.js";
import { createEmbeddedUiHandler, type EmbeddedAssets } from "./ui.js";
import { ArgsError, InitializationError } from "./errors.js";

/**
 * Run a Temps plugin. This is the main entry point.
 *
 * @example
 * ```ts
 * import { runPlugin, createManifest } from "@temps-sdk/plugin";
 *
 * runPlugin({
 *   manifest: () => createManifest("my-plugin", "0.1.0")
 *     .displayName("My Plugin")
 *     .addNav("Dashboard", "layout-dashboard", "/dashboard")
 *     .build(),
 *
 *   handler: (ctx) => async (req, res) => {
 *     res.writeHead(200, { "Content-Type": "application/json" });
 *     res.end(JSON.stringify({ hello: "world" }));
 *   },
 * });
 * ```
 */
export async function runPlugin(plugin: TempsPlugin): Promise<void> {
  // 1. Parse CLI arguments
  const args = parseArgs(process.argv.slice(2));

  // Ensure data directory exists
  mkdirSync(args.dataDir, { recursive: true });

  const manifest = plugin.manifest();

  // 2. Emit manifest (handshake phase 1)
  emitManifest(manifest);

  // Ensure socket directory exists and clean up stale socket
  const socketDir = dirname(args.socketPath);
  mkdirSync(socketDir, { recursive: true });
  if (existsSync(args.socketPath)) {
    unlinkSync(args.socketPath);
  }

  // Channel connection promise -- resolved when the host connects
  let resolveChannel: ((client: TempsClient) => void) | undefined;
  const channelReady = new Promise<TempsClient>((resolve) => {
    resolveChannel = resolve;
  });

  // Track state
  let currentContext: PluginContext | undefined;
  let pluginReady = false;

  // Embedded UI
  const embeddedAssets = plugin.embeddedUiAssets?.();
  const hasUi = embeddedAssets !== undefined || plugin.uiDistPath?.() !== undefined;

  // Build embedded UI handler if available
  const embeddedUiHandler = embeddedAssets
    ? createEmbeddedUiHandler(embeddedAssets)
    : undefined;

  // Active plugin fetch handler -- set once plugin is initialized
  let pluginFetch: ((req: Request) => Response | Promise<Response>) | undefined;

  // Bun WebSocket adapter plumbing
  const wsMessageListeners = new Map<unknown, Set<(data: string) => void>>();
  const wsCloseListeners = new Map<unknown, Set<() => void>>();

  // 3. Start Bun.serve on Unix socket
  const server = Bun.serve({
    unix: args.socketPath,

    fetch(req: Request, server) {
      const url = new URL(req.url);

      // WebSocket upgrade for platform channel
      if (url.pathname === PLUGIN_CHANNEL_PATH) {
        if (server.upgrade(req)) {
          return undefined as unknown as Response;
        }
        return new Response("WebSocket upgrade failed", { status: 500 });
      }

      // Health endpoint -- always available
      if (url.pathname === manifest.health_path || url.pathname === "/health") {
        return Response.json({ status: "ok", plugin: manifest.name });
      }

      // Event delivery endpoint
      if (url.pathname === "/_events" && req.method === "POST") {
        return handleEventDelivery(req, plugin, currentContext);
      }

      // Not initialized yet
      if (!pluginReady || !pluginFetch) {
        return new Response("Plugin initializing", { status: 503 });
      }

      // Try embedded UI
      if (embeddedUiHandler) {
        const uiResponse = handleEmbeddedUi(url, embeddedAssets!);
        if (uiResponse) return uiResponse;
      }

      // Delegate to plugin
      return pluginFetch(req);
    },

    websocket: {
      open(ws) {
        wsMessageListeners.set(ws, new Set());
        wsCloseListeners.set(ws, new Set());

        const adapter = createWsAdapter(ws, wsMessageListeners, wsCloseListeners);
        const client = new TempsClient(adapter);
        resolveChannel?.(client);
      },
      message(ws, message) {
        const msgStr = typeof message === "string" ? message : message.toString();
        const listeners = wsMessageListeners.get(ws);
        if (listeners) {
          for (const fn of listeners) fn(msgStr);
        }
      },
      close(ws) {
        const listeners = wsCloseListeners.get(ws);
        if (listeners) {
          for (const fn of listeners) fn();
        }
        wsMessageListeners.delete(ws);
        wsCloseListeners.delete(ws);
      },
    },
  });

  // 4. Emit ready (handshake phase 2)
  emitReady(hasUi);

  // 5. Wait for the host to connect via WebSocket channel (with timeout)
  const channelTimeoutMs = 30_000;
  let client: TempsClient;
  try {
    client = await Promise.race([
      channelReady,
      new Promise<never>((_, reject) =>
        setTimeout(
          () => reject(new Error("Channel connection timeout")),
          channelTimeoutMs
        )
      ),
    ]);
  } catch {
    console.error(
      "[temps-plugin] Warning: platform channel not connected, running without platform data access"
    );
    client = null as unknown as TempsClient;
  }

  // 6. Create PluginContext
  const ctx = new PluginContext({
    pluginName: manifest.name,
    dataDir: args.dataDir,
    authSecret: args.authSecret,
    client,
  });
  currentContext = ctx;

  if (client) {
    client.onEvent((event: PluginEvent) => {
      plugin.onEvent?.(ctx, event);
    });
  }

  // 7. Call plugin.onStart()
  try {
    await plugin.onStart?.(ctx);
  } catch (err) {
    throw new InitializationError(
      manifest.name,
      err instanceof Error ? err.message : String(err)
    );
  }

  // 8. Mount the plugin handler
  const nodeHandler = await plugin.handler(ctx);

  // Wrap the Node-style (req, res) handler into a Bun fetch handler
  pluginFetch = (req: Request) => nodeRequestBridge(req, nodeHandler);
  pluginReady = true;

  // 9. Graceful shutdown
  const shutdown = async () => {
    console.error(`[temps-plugin] Shutting down ${manifest.name}...`);
    try {
      await plugin.onShutdown?.();
    } catch (err) {
      console.error("[temps-plugin] Error during shutdown:", err);
    }
    client?.close();
    server.stop();

    if (existsSync(args.socketPath)) {
      try { unlinkSync(args.socketPath); } catch { /* ignore */ }
    }
    process.exit(0);
  };

  process.on("SIGTERM", shutdown);
  process.on("SIGINT", shutdown);

  console.error(
    `[temps-plugin] ${manifest.display_name ?? manifest.name} v${manifest.version} running on ${args.socketPath}`
  );
}

// ---------------------------------------------------------------------------
// Embedded UI serving (pure Bun Response)
// ---------------------------------------------------------------------------

function handleEmbeddedUi(
  url: URL,
  assets: EmbeddedAssets
): Response | undefined {
  const pathname = url.pathname;

  // Redirect /ui to /ui/
  if (pathname === "/ui") {
    return new Response(null, {
      status: 302,
      headers: { Location: "/ui/" },
    });
  }

  if (!pathname.startsWith("/ui/")) {
    return undefined;
  }

  const relativePath = pathname.slice("/ui/".length) || "index.html";

  if (relativePath.includes("..")) {
    return new Response("Forbidden", { status: 403 });
  }

  // Exact match
  const file = assets.get(relativePath);
  if (file) {
    return new Response(new Uint8Array(file.content), {
      headers: {
        "Content-Type": file.contentType,
        "Content-Length": String(file.content.byteLength),
        "Cache-Control": file.immutable
          ? "public, max-age=31536000, immutable"
          : "no-cache, no-store, must-revalidate",
      },
    });
  }

  // SPA fallback for paths without file extensions
  const hasExt = (relativePath.split("/").pop() ?? "").includes(".");
  if (!hasExt) {
    const index = assets.get("index.html");
    if (index) {
      return new Response(new Uint8Array(index.content), {
        headers: {
          "Content-Type": index.contentType,
          "Content-Length": String(index.content.byteLength),
          "Cache-Control": "no-cache, no-store, must-revalidate",
        },
      });
    }
  }

  return new Response("Not Found", { status: 404 });
}

// ---------------------------------------------------------------------------
// Bun WebSocket -> WsLike adapter
// ---------------------------------------------------------------------------

function createWsAdapter(
  ws: { send: (msg: string | Buffer) => void; close: () => void },
  messageListeners: Map<unknown, Set<(data: string) => void>>,
  closeListeners: Map<unknown, Set<() => void>>
): WsLike {
  return {
    on(event: string, fn: (...args: unknown[]) => void) {
      if (event === "message") {
        messageListeners.get(ws)?.add(fn as (data: string) => void);
      } else if (event === "close") {
        closeListeners.get(ws)?.add(fn as () => void);
      }
    },
    send(data: string, cb?: (err?: Error) => void) {
      try {
        ws.send(data);
        cb?.();
      } catch (err) {
        cb?.(err as Error);
      }
    },
    close() {
      ws.close();
    },
  };
}

// ---------------------------------------------------------------------------
// Node (req, res) handler -> Bun fetch bridge
// ---------------------------------------------------------------------------

import type { IncomingMessage, ServerResponse } from "node:http";

async function nodeRequestBridge(
  req: Request,
  handler: (req: IncomingMessage, res: ServerResponse) => void | Promise<void>
): Promise<Response> {
  const url = new URL(req.url);

  // Read body upfront
  let bodyBuffer = Buffer.alloc(0);
  if (req.body) {
    bodyBuffer = Buffer.from(await req.arrayBuffer());
  }

  return new Promise<Response>((resolve) => {
    // Build a minimal IncomingMessage shim
    const fakeReq = Object.create(null) as IncomingMessage;
    fakeReq.url = url.pathname + url.search;
    fakeReq.method = req.method;

    const hdrs: Record<string, string> = {};
    req.headers.forEach((v, k) => { hdrs[k] = v; });
    fakeReq.headers = hdrs;

    // Body event emitter shim
    const dataFns: Array<(chunk: Buffer) => void> = [];
    const endFns: Array<() => void> = [];
    fakeReq.on = ((event: string, fn: (...a: unknown[]) => void) => {
      if (event === "data") dataFns.push(fn as (c: Buffer) => void);
      if (event === "end") endFns.push(fn as () => void);
      return fakeReq;
    }) as typeof fakeReq.on;

    // Response collector shim
    let statusCode = 200;
    const resHeaders: Record<string, string> = {};
    const chunks: Buffer[] = [];

    const fakeRes = Object.create(null) as ServerResponse;
    fakeRes.writeHead = ((code: number, h?: Record<string, string>) => {
      statusCode = code;
      if (h) Object.assign(resHeaders, h);
      return fakeRes;
    }) as typeof fakeRes.writeHead;

    fakeRes.end = ((data?: string | Buffer) => {
      if (data) chunks.push(typeof data === "string" ? Buffer.from(data) : data);
      const body = Buffer.concat(chunks);
      resolve(new Response(body.length > 0 ? body : null, {
        status: statusCode,
        headers: resHeaders,
      }));
    }) as typeof fakeRes.end;

    fakeRes.write = ((data: string | Buffer) => {
      chunks.push(typeof data === "string" ? Buffer.from(data) : data);
      return true;
    }) as typeof fakeRes.write;

    // Call the handler
    handler(fakeReq, fakeRes);

    // Deliver body
    queueMicrotask(() => {
      if (bodyBuffer.length > 0) {
        for (const fn of dataFns) fn(bodyBuffer);
      }
      for (const fn of endFns) fn();
    });
  });
}

// ---------------------------------------------------------------------------
// Event delivery handler (POST /_events)
// ---------------------------------------------------------------------------

async function handleEventDelivery(
  req: Request,
  plugin: TempsPlugin,
  ctx?: PluginContext
): Promise<Response> {
  try {
    const event = (await req.json()) as PluginEvent;
    if (ctx) {
      plugin.onEvent?.(ctx, event);
    }
    return Response.json({ ok: true });
  } catch (err) {
    return Response.json(
      {
        error: "Invalid event payload",
        detail: err instanceof Error ? err.message : String(err),
      },
      { status: 400 }
    );
  }
}

// ---------------------------------------------------------------------------
// CLI argument parsing
// ---------------------------------------------------------------------------

function parseArgs(argv: string[]): PluginArgs {
  const args: Partial<PluginArgs> = {};

  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i]!;
    const next = argv[i + 1];

    switch (arg) {
      case "--socket-path":
        args.socketPath = next;
        i++;
        break;
      case "--database-url":
        args.databaseUrl = next;
        i++;
        break;
      case "--auth-secret":
        args.authSecret = next;
        i++;
        break;
      case "--data-dir":
        args.dataDir = next;
        i++;
        break;
    }
  }

  if (!args.socketPath) {
    throw new ArgsError("--socket-path is required");
  }
  if (!args.authSecret) {
    throw new ArgsError("--auth-secret is required");
  }
  if (!args.dataDir) {
    throw new ArgsError("--data-dir is required");
  }

  return args as PluginArgs;
}
